src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,323 @@
+                greenbidsAnalyticsProperties.getPbuid(),
+                greenbidsAnalyticsProperties.getGreenbidsSampling()
+        );
+        this.jacksonMapper = Objects.requireNonNull(jacksonMapper);
+        this.isSampled = null;
+        this.greenbidsId = null;
+        this.billingId = null;
+    }
+
+    public CommonMessage createCommonMessage(AuctionContext auctionContext, Long auctionElapsed) {
+        return new CommonMessage(
+                auctionContext,
+                greenbidsConfig,
+                auctionElapsed,
+                greenbidsId,
+                billingId
+        );
+    }
+
+    private String getGpid(ObjectNode impExt) {
+        return Optional.ofNullable(impExt)
+                .map(ext -> ext.get("prebid"))
+                .map(JsonNode::asText)
+                .orElse(null);
+    }
+
+    private ExtImpPrebid extImpPrebid(JsonNode extImpPrebid) {
+        try {
+            return jacksonMapper.mapper().treeToValue(extImpPrebid, ExtImpPrebid.class);
+        } catch (JsonProcessingException e) {
+            throw new PreBidException("Error decoding imp.ext.prebid: " + e.getMessage(), e);
+        }
+    }
+
+    private String storedRequestId(ObjectNode impExt) {
+        return Optional.ofNullable(impExt)
+                .map(ext -> ext.get("prebid"))
+                .map(this::extImpPrebid)
+                .map(ExtImpPrebid::getStoredrequest)
+                .map(ExtStoredRequest::getId)
+                .orElse(null);
+    }
+
+    private String getAdUnitCode(Imp imp) {
+        final ObjectNode impExt = imp.getExt();
+
+        final String adUnitCodeStoredRequestId = Optional.ofNullable(getGpid(impExt))
+                .orElseGet(() -> storedRequestId(impExt));
+
+        if (adUnitCodeStoredRequestId != null) {
+            return adUnitCodeStoredRequestId;
+        }
+
+        return imp.getId();
+    }
+
+    public CommonMessage addBidResponseToMessage(
+            CommonMessage commonMessage,
+            List<Imp> imps,
+            Map<String, Bid> seatsWithBids,
+            Map<String, NonBid> seatsWithNonBids
+    ) {
+        commonMessage.adUnits = imps.stream().map(imp -> {
+            final Banner banner = imp.getBanner();
+            final Video video = imp.getVideo();
+            final Native nativeObject = imp.getXNative();
+
+            final Integer width = banner.getFormat().get(0).getW();
+            final Integer height = banner.getFormat().get(0).getH();
+
+            final ExtBanner extBanner = ExtBanner.builder()
+                    .sizes(
+                            width != null && height != null ? Arrays.asList(
+                                    Arrays.asList(width, height),
+                                    Arrays.asList(width, height)
+                            ) : null
+                    )
+                    .pos(banner.getPos())
+                    .name(banner.getId())
+                    .build();
```

Comment:
I've rebuild the Banner into ExtBanner (that contains `List<List<Integer>> sizes`) because in the our current implem of [_Banner](https://github.com/greenbids/data-processing/blob/a6d82b424a321ea003fb63b4d416e3e0e1008ecd/src/greenbids/streaming/data_types.py#L38-L39) in data-processing the sizes field is `list[list[int]]` 

In the implem of [Banner](https://github.com/EvgeniiMunin/prebid-server-java/blob/338f24c0d45f8ae9809f7f3d3eb143581ece9fcf/src/main/java/com/iab/openrtb/request/Banner.java#L26) type in PBS it has a List<Format> format  where each [Format](https://github.com/EvgeniiMunin/prebid-server-java/blob/338f24c0d45f8ae9809f7f3d3eb143581ece9fcf/src/main/java/com/iab/openrtb/request/Format.java#L18-L26) contains width and height. In the list of formats their sizes can be different as well

ex:
```
"format": [{
        "w": 600,
        "h": 500
      }, {
        "w": 300,
        "h": 600
      }]
```

Reply:
After looking at whether we can cast List[List[Integer]] from List[Format] I think the following


- Banner contains List[Format] and Format contains w, h 
- sizes is mandatory field in analytics server schema
- so I need to create custom class ExtBanner that cast List[List[Integer]] -> List[Format]
	- implement the same fields in the ExtBanner class
	- or inherit Banner class. But I cannot inherit from Banner class because it is final

Reply:
[fixed](https://github.com/prebid/prebid-server-java/pull/3096/commits/c2b2a20d53a74a13ac388dc03dca3a9383291175) extraction of sizes width height from Format object , but still need to use ExtBanner class